salvo/src/lib/benchmark/fully_dockerized_benchmark.py

```diff
@@ -24,25 +24,132 @@ def __init__(self, job_control: JobControl, benchmark_name: str) -> None:
     super(Benchmark, self).__init__(job_control, benchmark_name)
 
   def _validate(self) -> None:
-    """Validate that all data required for running a benchmark exist.
+    """Validate that all data required for running a benchmark exists.
+
+    Verify that all required images are present in the control object.
+    If not, verify that sources exist from which we can build the
+    required docker images.
 
     Returns:
         None
     """
-    pass
+    verify_source = False
+    images = self.get_images()
+
+    # Determine whether we need to build the images from source
+    # If so, verify that the required source data is defined
+    verify_source = images is None or \
+        not images.nighthawk_benchmark_image or \
+        not images.nighthawk_binary_image or \
+        not images.envoy_image
+
+    log.debug(f"Source verification needed: {verify_source}")
+    if verify_source:
+      self._verify_sources(images)
+
+    return
 
   def _verify_sources(self, images: DockerImages) -> None:
     """Validate that sources are available to build a missing image.
 
+    Verify that a source definition exists tht can build a missing
+    image needed for the benchmark.
+
     Returns:
         None
+
+    Raises:
+        an Exception if no source definitions allow us to build
+          missing docker images.
     """
-    pass
+    source = self.get_source()
+    if not source:
+      raise Exception("No source configuration specified")
+
+    can_build_envoy = False
+    can_build_nighthawk = False
+
+    for source_def in source:
+      # Cases:
+      # Missing envoy image -> Need to see an envoy source definition
+      # Missing at least one nighthawk image -> Need to see a nighthawk source
+
+      if source_def.identity == source_def.SRCID_UNSPECIFIED:
+        raise Exception("No source identity specified")
+
+      if not images.envoy_image \
+          and source_def.identity == source_def.SRCID_ENVOY and \
+          (source_def.source_path or source_def.source_url):
+        can_build_envoy = True
+
+      if (not images.nighthawk_benchmark_image or not images.nighthawk_binary_image) \
+          and source_def.identity == source_def.SRCID_NIGHTHAWK and \
+          (source_def.source_path or source_def.source_url):
+        can_build_nighthawk = True
+
+      if (not images.envoy_image and not can_build_envoy) or \
+          (not images.nighthawk_benchmark_image or not images.nighthawk_binary_image) \
+              and not can_build_nighthawk:
+        # If the Envoy image is specified, then the validation failed for
+        # NightHawk and vice versa
+        msg = "No source specified to build unspecified {image} image".format(
```

Comment:
nit: Can we spell out the full name for the variable, i.e. `message` instead of just `msg`? The style guide advises against abbreviations:

https://github.com/google/styleguide/blob/gh-pages/pyguide.md#316-naming

Reply:
Certainly, this isn't javascript after all :)  Will update.